google/cloud/forseti/services/inventory/storage.py

```diff
@@ -730,13 +713,26 @@ def write(self, resource):
         Raises:
             Exception: If storage was opened readonly.
         """
-
         if self.readonly:
             raise Exception('Opened storage readonly')
 
+        previous_id = self._get_resource_id(resource)
+        if previous_id:
+            resource.set_inventory_key(previous_id)
+            self.update(resource)
+            return
+
         rows = Inventory.from_resource(self.inventory_index, resource)
+
+        # Insert the resource row to get the inventory id, buffer other rows.
         for row in rows:
-            self.buffer.add(row)
+            if row.category == Categories.resource:
+                self.session.add(row)
+                self.session.flush()
```

Comment:
I think flushing everytime is very expensive, using the buffer should be able to increase the performance, I tried flushing every row vs the db buffer flush every 1024 rows, the difference is roughly 1 minute in our test domain (3mins 2x seconds vs 4mins 2x seconds)

Reply:
I made this more explicit in the code. If you re-run the test with the new schema, you'll see this code is significantly faster than the previous version due to the more efficient schema and index.